### PR TITLE
PythonPackage: check purelib for libs/headers

### LINKED ISSUE #42441: Installation issue: py-matplotlib
### Steps to reproduce the issue

```console
$ spack spec -I <spec>
Input spec
--------------------------------
 -   py-matplotlib

Concretized
--------------------------------
 -   py-matplotlib@3.8.2%apple-clang@15.0.0~animation~fonts~latex~movies backend=macosx build_system=python_pip arch=darwin-ventura-m1
[+]      ^freetype@2.11.1%apple-clang@15.0.0+pic+shared build_system=autotools arch=darwin-ventura-m1
[+]          ^bzip2@1.0.8%apple-clang@15.0.0~debug~pic+shared build_system=generic arch=darwin-ventura-m1
[+]              ^diffutils@3.9%apple-clang@15.0.0 build_system=autotools arch=darwin-ventura-m1
[+]                  ^libiconv@1.17%apple-clang@15.0.0 build_system=autotools libs=shared,static arch=darwin-ventura-m1
[e]          ^gmake@3.81%apple-clang@15.0.0~guile build_system=generic patches=ca60bd9 arch=darwin-ventura-m1
[+]          ^gnuconfig@2022-09-17%apple-clang@15.0.0 build_system=generic arch=darwin-ventura-m1
[+]      ^libpng@1.6.39%apple-clang@15.0.0~ipo~pic build_system=cmake build_type=Release generator=make libs=shared,static arch=darwin-ventura-m1
[+]          ^cmake@3.27.9%apple-clang@15.0.0~doc+ncurses+ownlibs build_system=generic build_type=Release arch=darwin-ventura-m1
[+]              ^curl@8.4.0%apple-clang@15.0.0~gssapi~ldap~libidn2~librtmp~libssh~libssh2+nghttp2 build_system=autotools libs=shared,static tls=secure_transport arch=darwin-ventura-m1
[+]                  ^nghttp2@1.57.0%apple-clang@15.0.0 build_system=autotools arch=darwin-ventura-m1
[+]              ^ncurses@6.4%apple-clang@15.0.0~symlinks+termlib abi=none build_system=autotools arch=darwin-ventura-m1
[+]          ^zlib-ng@2.1.5%apple-clang@15.0.0+compat+opt build_system=autotools arch=darwin-ventura-m1
[+]      ^pkgconf@1.9.5%apple-clang@15.0.0 build_system=autotools arch=darwin-ventura-m1
[+]      ^py-certifi@2023.7.22%apple-clang@15.0.0 build_system=python_pip arch=darwin-ventura-m1
[+]      ^py-contourpy@1.0.7%apple-clang@15.0.0 build_system=python_pip arch=darwin-ventura-m1
[+]      ^py-cycler@0.11.0%apple-clang@15.0.0 build_system=python_pip arch=darwin-ventura-m1
[+]      ^py-fonttools@4.39.4%apple-clang@15.0.0 build_system=python_pip arch=darwin-ventura-m1
[+]      ^py-importlib-resources@5.12.0%apple-clang@15.0.0 build_system=python_pip arch=darwin-ventura-m1
[+]          ^py-zipp@3.17.0%apple-clang@15.0.0 build_system=python_pip arch=darwin-ventura-m1
[+]      ^py-kiwisolver@1.4.5%apple-clang@15.0.0 build_system=python_pip arch=darwin-ventura-m1
[+]          ^py-cppy@1.2.1%apple-clang@15.0.0 build_system=python_pip arch=darwin-ventura-m1
[+]      ^py-numpy@1.26.3%apple-clang@15.0.0 build_system=python_pip patches=873745d arch=darwin-ventura-m1
[+]          ^ninja@1.11.1%apple-clang@15.0.0+re2c build_system=generic arch=darwin-ventura-m1
[+]              ^re2c@2.2%apple-clang@15.0.0 build_system=generic arch=darwin-ventura-m1
[+]          ^openblas@0.3.26%apple-clang@15.0.0~bignuma~consistent_fpcsr+fortran~ilp64+locking+pic+shared build_system=makefile symbol_suffix=none threads=none arch=darwin-ventura-m1
[e]              ^perl@5.34.1%apple-clang@15.0.0~cpanm+opcode+open+shared+threads build_system=generic arch=darwin-ventura-m1
[+]          ^py-cython@3.0.8%apple-clang@15.0.0 build_system=python_pip arch=darwin-ventura-m1
[+]          ^py-pyproject-metadata@0.7.1%apple-clang@15.0.0 build_system=python_pip arch=darwin-ventura-m1
[+]          ^py-tomli@2.0.1%apple-clang@15.0.0 build_system=python_pip arch=darwin-ventura-m1
[+]      ^py-packaging@23.1%apple-clang@15.0.0 build_system=python_pip arch=darwin-ventura-m1
[+]          ^py-flit-core@3.9.0%apple-clang@15.0.0 build_system=python_pip arch=darwin-ventura-m1
[+]      ^py-pillow@10.0.0%apple-clang@15.0.0~freetype~imagequant+jpeg~jpeg2000~lcms~raqm~tiff~webp~xcb+zlib build_system=python_pip arch=darwin-ventura-m1
[+]          ^libjpeg-turbo@3.0.0%apple-clang@15.0.0~ipo~jpeg8~partial_decoder+shared+static build_system=cmake build_type=Release generator=make arch=darwin-ventura-m1
[+]              ^nasm@2.15.05%apple-clang@15.0.0 build_system=autotools arch=darwin-ventura-m1
[+]      ^py-pip@23.1.2%apple-clang@15.0.0 build_system=generic arch=darwin-ventura-m1
[+]      ^py-pybind11@2.11.1%apple-clang@15.0.0~ipo build_system=cmake build_type=Release generator=ninja arch=darwin-ventura-m1
[+]      ^py-pyparsing@3.0.9%apple-clang@15.0.0 build_system=python_pip arch=darwin-ventura-m1
[+]      ^py-python-dateutil@2.8.2%apple-clang@15.0.0 build_system=python_pip arch=darwin-ventura-m1
[+]          ^py-six@1.16.0%apple-clang@15.0.0 build_system=python_pip arch=darwin-ventura-m1
[+]      ^py-setuptools@68.0.0%apple-clang@15.0.0 build_system=generic arch=darwin-ventura-m1
[+]      ^py-setuptools-scm@7.1.0%apple-clang@15.0.0+toml build_system=python_pip arch=darwin-ventura-m1
[e]          ^git@2.39.3%apple-clang@15.0.0+man+nls+perl+subtree~svn~tcltk build_system=autotools arch=darwin-ventura-m1
[+]          ^py-typing-extensions@4.8.0%apple-clang@15.0.0 build_system=python_pip arch=darwin-ventura-m1
[+]      ^py-wheel@0.41.2%apple-clang@15.0.0 build_system=generic arch=darwin-ventura-m1
[e]      ^python@3.9.6%apple-clang@15.0.0+bz2+crypt+ctypes+dbm~debug+libxml2+lzma+nis~optimizations+pic+pyexpat~pythoncmd+readline+shared+sqlite3+ssl+tix+tkinter+uuid+zlib build_system=generic patches=0d98e93,ebdca64,f2fd060 arch=darwin-ventura-m1
[+]      ^qhull@2020.2%apple-clang@15.0.0~ipo build_system=cmake build_type=Release generator=make arch=darwin-ventura-m1
```

### Error message

<details><summary>raise NoHeadersError</summary>

<pre>
    raise NoHeadersError(msg.format(self.spec.name, include, platlib))
spack.error.NoHeadersError: Unable to locate py-numpy headers in /Users/srinath22/REPOS/Github/spack/opt/spack/darwin-ventura-m1/apple-clang-15.0.0/py-numpy-1.26.3-xrkzszwugmtpy4nanwppleujd2bkyigw/include/python3.9/numpy or /Users/srinath22/REPOS/Github/spack/opt/spack/darwin-ventura-m1/apple-clang-15.0.0/py-numpy-1.26.3-xrkzszwugmtpy4nanwppleujd2bkyigw/lib64/python3.9/site-packages/numpy
</pre></details>

### Information on your system

* **Spack:** 0.22.0.dev0 (5ef9bb775276c32e3a98b4f268a29d020e514c77)
* **Python:** 3.9.6
* **Platform:** darwin-ventura-m1
* **Concretizer:** clingo

### Additional information

@adamjstewart  Greg @becker33 pointed me to the platlib function in `python/package.py` calculating the incorrect location of py-numpy header files (some where near line 1094). We did an exercise of using `spack python` and it showed that the path calculated did not match py-numpy installed package. 

The stage directory is empy. So no `spack-build-*` files.   This error happens IN and OUT of an enviroment. 

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

### PR DESCRIPTION
**Pull request by adamjstewart**
Fixes #42441, where the libs/headers somehow ended up in purelib instead of platlib.

Let me know if the functional programming stuff is hard to read. I can replace it with for-loops if needed.

@srinathv

### CODE REVIEW

**Review by rgommers — APPROVED** [review]
This looks fine to me, if CI is happy then it should be okay to merge I think.

**Review by becker33 — APPROVED** [review]
My only concern is performance on externals in large directories.

**Comment by adamjstewart** [comment]
> My only concern is performance on externals in large directories.

This shouldn't be an issue. We always search in the `lib/pythonX.Y/site-packages/foo` and `include/foo` directories for package `foo`, we don't search in their parent directories.

**Comment by wdconinc** [comment]
This PR introduced a regression, and on a `spack -d install` I'm getting:
```
Process Process-1:
Traceback (most recent call last):
  File "/home/wdconinc/git/spack/lib/spack/spack/build_environment.py", line 1112, in _setup_pkg_and_run
    kwargs["env_modifications"] = setup_package(
                                  ^^^^^^^^^^^^^^
  File "/home/wdconinc/git/spack/lib/spack/spack/build_environment.py", line 754, in setup_package
    set_wrapper_variables(pkg, env_mods)
  File "/home/wdconinc/git/spack/lib/spack/spack/build_environment.py", line 534, in set_wrapper_variables
    update_compiler_args_for_dep(dspec)
  File "/home/wdconinc/git/spack/lib/spack/spack/build_environment.py", line 521, in update_compiler_args_for_dep
    _prepend_all(include_dirs, query.headers.directories)
                               ^^^^^^^^^^^^^
  File "/home/wdconinc/git/spack/lib/spack/spack/spec.py", line 1227, in __get__
    value = f()
            ^^^
  File "/home/wdconinc/git/spack/lib/spack/spack/spec.py", line 1216, in <lambda>
    callbacks_chain.append(lambda: getattr(pkg, self.attribute_name))
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/wdconinc/git/spack/lib/spack/spack/build_systems/python.py", line 381, in headers
    headers = functools.reduce(operator.add, headers_list)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/wdconinc/git/spack/lib/spack/llnl/path.py", line 99, in path_filter_caller
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
TypeError: find_all_headers() got an unexpected keyword argument 'recursive'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.11/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.11/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/wdconinc/git/spack/lib/spack/spack/build_environment.py", line 1132, in _setup_pkg_and_run
    package_context = get_package_context(tb)
                      ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/wdconinc/git/spack/lib/spack/spack/build_environment.py", line 1340, in get_package_context
    func = getattr(obj, tb.tb_frame.f_code.co_name, "")
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/wdconinc/git/spack/lib/spack/spack/build_systems/python.py", line 381, in headers
    headers = functools.reduce(operator.add, headers_list)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/wdconinc/git/spack/lib/spack/llnl/path.py", line 99, in path_filter_caller
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
TypeError: find_all_headers() got an unexpected keyword argument 'recursive'
```

Posting here since it's caused by this PR: reverting fixes it. I will investigate further and post an issue.

File: lib/spack/spack/build_systems/python.py [hunk 1]
```diff
@@ -346,16 +349,20 @@ def headers(self) -> HeaderList:
...
         name = self.spec.name[3:]
 
-        # Headers may be in either location
+        # Headers should only be in include or platlib, but no harm in checking purelib too
         include = self.prefix.join(self.spec["python"].package.include).join(name)
         platlib = self.prefix.join(self.spec["python"].package.platlib).join(name)
-        headers = fs.find_all_headers(include) + fs.find_all_headers(platlib)
+        purelib = self.prefix.join(self.spec["python"].package.purelib).join(name)
+
+        find_all_headers = functools.partial(fs.find_all_headers, recursive=True)
```
Comment (alalazo) [inline lib/spack/spack/build_systems/python.py:357]: @wdconinc There is an issue here. `find_all_headers` has a different signature than `find all_*_libraries` :sweat: and takes no `recursive=...` argument.

Comment (alalazo) [reply lib/spack/spack/build_systems/python.py:357]: Let me submit a quick fix

**Pull request merged**